Fixes bgp ranking#1901

g4ze opened the pull request:
closes https://github.com/intelowlproject/IntelOwl/issues/1901

# Description

added new analyzer BGP_ranking
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [x] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [x] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [x] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [x] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [ ] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

![image](https://github.com/intelowlproject/IntelOwl/assets/102874321/51800bc1-e754-42cf-bb50-bc0b40c5026d)
{
    "asn": "26615",
    "asn_rank": 0,
    "asn_position": None,
    "asn_description": "TIM SA, BR"
}

0ssigeno commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+
+    def run(self):
+        logger.info("Running BGP_Ranking")
+        urls = {
+            "getASN": "https://bgpranking-ng.circl.lu/ipasn_history/?ip=",
+            "getASNRank": "https://bgpranking-ng.circl.lu/json/asn",
+            "getASNHistory": "https://bgpranking-ng.circl.lu/json/asn_history",
+        }
```
Can we set these as class attributes? They don't need to be calculated at every execution

g4ze replied:
oh yeah, right. lemme fix that.

0ssigeno commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+            finalresposne["asn"] = response["response"][
+                list(response["response"].keys())[0]
+            ]["asn"]
+            logger.info("ASN extracted from IP")
+
+            # get ASN rank from extracted ASN
+
+            response = requests.post(
+                urls["getASNRank"], data=json.dumps({"asn": finalresposne["asn"]})
+            )
+            response.raise_for_status()
+            response = response.json()
+            finalresposne["asn_description"] = response["response"]["asn_description"]
+            finalresposne["asn_rank"] = response["response"]["ranking"]["rank"]
+            finalresposne["asn_position"] = response["response"]["ranking"]["position"]
+            logger.info("ASN rank and position extracted from ASN")
+
+            if self.period:
+                # get ASN history from extracted ASN
+                response = requests.post(
+                    urls["getASNHistory"],
+                    data=json.dumps(
+                        {"asn": finalresposne["asn"], "period": self.period}
+                    ),
+                )
+                response.raise_for_status()
+                response = response.json()
+                finalresposne["asn_history"] = response["response"]["asn_history"]
+                logger.info("ASN history extracted from ASN")
+
+        except (
+            requests.exceptions.RequestException,
+            TypeError,
+            json.JSONDecodeError,
+            KeyError,
+            AttributeError,
+        ) as e:
+            # Handle various specific exceptions
+            logger.error(f"Exception: {e}")
+            raise AnalyzerRunException(f"AnalyzerRunException: {e}")
```
I **hate** a try catch that big that pretty much manage every single line of execution.
Moreover, if in the catch branch we are not doing actually anything.
Imho we can just remove the try catch completely and let the `start` function manage this accordingly

g4ze replied:
okay, I see. Please correct me if I'm wrong here, I need to get rid of the error handling in the module entirely, right? by start function you mean `run` right?
> Moreover, if in the catch branch we are not doing actually anything.
Is it okay to not log and raise an exception? I'm a bit confused here...

0ssigeno replied:
> I need to get rid of the error handling in the module entirely, right?
right
> by start function you mean run right?
Nope, i mean the `start` function. Its code is in the `plugin` parent class.
>Is it okay to not log and raise an exception? I'm a bit confused here...
If you check the code of the `start` function you will see that we are already doing exactly that(and else) in the `after_run_failed` error handler function

g4ze replied:
done.

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+            finalresposne["asn"] = response["response"][
+                list(response["response"].keys())[0]
+            ]["asn"]
+            logger.info("ASN extracted from IP")
```
static string logs are almost useless. Please add content otherwise it does not help while debugging the app

mlodic replied:
plus add a log also _before_ the request plz

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+            response = requests.get(urls["getASN"] + self.observable_name)
+            response.raise_for_status()
+            response = response.json()
+            finalresposne["asn"] = response["response"][
+                list(response["response"].keys())[0]
+            ]["asn"]
```
simplify this line and manage existence of the keys accordingly here and not at the end

g4ze replied:
```asn = response.get("response", {}).popitem()[1].get("asn", None)
            if not asn:
                raise AnalyzerRunException("ASN not found in response")```
something like this?

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+            finalresposne["asn_description"] = response["response"]["asn_description"]
+            finalresposne["asn_rank"] = response["response"]["ranking"]["rank"]
+            finalresposne["asn_position"] = response["response"]["ranking"]["position"]
+            logger.info("ASN rank and position extracted from ASN")
```
same here

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+                response.raise_for_status()
+                response = response.json()
+                finalresposne["asn_history"] = response["response"]["asn_history"]
+                logger.info("ASN history extracted from ASN")
```
and here

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,102 @@
+            },
+        }
+
+        patches = [
+            if_mock_connections(
+                patch(
+                    "requests.get",
+                    return_value=MockUpResponse(response, 200),
+                ),
+            )
+        ]
```
you could also patch the additional requests (cause we make 3 requests in this analyzer). See vt3_get.py file as an example

g4ze replied:
done

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,143 @@
+                )
+                if final_response["asn_history"] is None:
+                    raise AnalyzerRunException(f"ASN history not found in {response}")
+                logger.info(f"ASN history: {final_response['asn_history']}")
```
please add `{self.observable_name}` in the log so we can find related logs

mlodic commented on api_app/analyzers_manager/migrations/0069_analyzer_config_bgp_ranking.py:
```diff
@@ -0,0 +1,152 @@
+        },
+        "name": "period",
+        "type": "int",
+        "description": "period for bgp-ranking",
```
from the description is not clear which kind of period. Hours/minutes/days?

g4ze commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,143 @@
+    wrapper for https://github.com/D4-project/BGP-Ranking
+    """
+
+    getASN: str = "https://bgpranking-ng.circl.lu/ipasn_history/?ip="
+    getASNRank: str = "https://bgpranking-ng.circl.lu/json/asn"
+    getASNHistory: str = "https://bgpranking-ng.circl.lu/json/asn_history"
+    observable_name: str
```
removing extra variables

g4ze commented:
@mlodic kindly review

mlodic commented on api_app/analyzers_manager/migrations/0069_analyzer_config_bgp_ranking.py:
```diff
@@ -0,0 +1,152 @@
+    "routing_key": "default",
+    "health_check_status": True,
+    "type": "observable",
+    "docker_based": False,
```
other worries came up to my mind. Can you please add a `health_check` schedule here? considering that this service could be down and is not guaranteed, it would make sense to check it once every hour for instance.

that should be something that you already configured on other analyzers if I remember correctly

mlodic replied:
we need to have a parameter called either `url` or `_url` to have the default health check works. See: https://github.com/intelowlproject/IntelOwl/blob/master/api_app/classes.py#L239

mlodic replied:
your changes do not reflect what I asked here

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,144 @@
+        # get ASN from ip
+
+        logger.info(f"Extracting ASN from IP: {self.observable_name}")
+        response = requests.get(
+            self.base_url + "/ipasn_history/?ip=" + self.observable_name
+        )
+        response.raise_for_status()
```
I would like to add another parameter that we usually add when interacting with free services: `timeout`.
That would be the timeout for the requests. If we don't set it, these requests can keep going for a long time and this analyzer can become the bottleneck of a playbook. This could be really a big problem.

The default should be 5 seconds, like in other analyzers. Thank you

mlodic reviewed (changes requested):
almost there, just 2 other things to help managing the analyze service

g4ze commented:
Hi! @mlodic was a bit bz with college and a side project, hence the delay. Kindly review.

mlodic commented on api_app/analyzers_manager/observable_analyzers/bgp_ranking.py:
```diff
@@ -0,0 +1,159 @@
+
+        return final_response
+
+    def health_check(self) -> bool:
+        """
+        basic health check: if instance is up or not
+        """
+        try:
+            requests.head(self.base_url, timeout=self.timeout)
+        except requests.exceptions.RequestException:
+            return False
+
+        return True
```
this is not required, the framework already does this once you set up a health check schedule

g4ze replied:
My bad, overlooked a comment.

g4ze commented:
can you please tell me if you want 
```
"health_check_schedule": {
            "minute": "0",
            "hour": "0",
            "day_of_week": "*",
            "day_of_month": "*",
            "month_of_year": "*",
        },
```
outside the `python_module` key as there doesn't seem to be any predefined health_check schedule set to null.
Do you want me to add a new key and value there?

mlodic commented:
honestly I didn't understand your last comment.

The cron time period is wrong cause that schedule would work once a day, not once an hour

g4ze commented:
okay so, I'm trying to say that I've already added the cron `health_check_schedule` after you asked for it, inside the `plugin` dictionary's `python module` parameter(which was once a day, changed now to once a hour). 
The part where you left the comment to add the schedule does not have a paramenter for `health_check_schedule`. The parameter was inside `python_module` key of the `plugin` dictionary, where I added the schedule.
My question is if you want me to add the schedule in the place where you exactly commented (which is outside the `python_module` parameter of `plugin` dictionary )

mlodic commented:
ah I got your doubt now. Now, that was not intended, it is right how you did it!

mlodic commented:
great addition this analyzer! :)

g4ze commented:
It's so weirdly satisfying haha ◖⁠⚆⁠ᴥ⁠⚆⁠◗
